app/src/heartbeat/heartbeat_controller_desktop.cc

```diff
@@ -79,5 +97,171 @@ void HeartbeatController::LogHeartbeat() {
+        }
+      }
+    }
+    // Post the semaphore to signal the main thread to read the result.
+    scheduled_work_semaphore->Post();
+    return;
+  };
+
+  scheduler_.Schedule(get_and_reset_function);
+  // Wait until the scheduled work completes.
+  if (scheduled_work_semaphore->TimedWait(kMaxWaitTimeMs)) {
+    return *output_str;
+  }
+  // Return an empty string if unable to TimedWait times out.
+  return "";
+}
+
+std::string HeartbeatController::GetStringPayloadForHeartbeats(
+    LoggedHeartbeats heartbeats) {
+  Variant heartbeats_variant = Variant::EmptyVector();
+  std::vector<Variant>& heartbeats_vector = heartbeats_variant.vector();
+
+  // heartbeats is a map from user agent to a vector of date strings.
+  for (auto const& entry : heartbeats.heartbeats) {
+    Variant heartbeat_entry = Variant::EmptyMap();
+    std::map<Variant, Variant>& heartbeat_entry_map = heartbeat_entry.map();
+    Variant dates_variant = Variant::EmptyVector();
+    std::vector<Variant>& dates_vector = dates_variant.vector();
+    std::string user_agent = entry.first;
+    for (std::string date : entry.second) {
+      dates_vector.push_back(date);
+    }
+    heartbeat_entry_map[kUserAgentKey] = user_agent;
+    heartbeat_entry_map[kDatesKey] = dates_variant;
+    heartbeats_vector.push_back(heartbeat_entry);
+  }
+  // Construct a final object with 'heartbeats' and 'version'
+  Variant root = Variant::EmptyMap();
+  std::map<Variant, Variant>& root_map = root.map();
+  root_map[kHeartbeatsKey] = heartbeats_variant;
+  root_map[kVersionKey] = "2";
+
+  std::string json_object = util::VariantToJson(root);
+  return CompressAndEncode(json_object);
+}
+
+std::string HeartbeatController::CompressAndEncode(const std::string& input) {
+  ZLib zlib;
+  zlib.SetGzipHeaderMode();
+  uLongf result_size = ZLib::MinCompressbufSize(input.length());
+  std::unique_ptr<char[]> result(new char[result_size]);
+  int err = zlib.Compress(
+      reinterpret_cast<unsigned char*>(result.get()), &result_size,
+      reinterpret_cast<const unsigned char*>(input.data()), input.length());
+  if (err != Z_OK) {
+    // Failed to compress.
+    return "";
+  }
+  std::string compressed_str = std::string(result.get(), result_size);
+
+  std::string output;
+  if (!firebase::internal::Base64EncodeUrlSafe(compressed_str, &output)) {
+    // Failed to based64 encode.
+    return "";
+  }
+  // If compressed payload is too large, return an empty string instead.
+  if (output.size() > kMaxPayloadSize) {
+    return "";
+  }
+  return output;
+}
+
+std::string HeartbeatController::DecodeAndDecompress(const std::string& input) {
+  std::string decoded;
+  if (!firebase::internal::Base64Decode(input, &decoded)) {
+    // Failed to decode.
+    return "";
+  }
+
+  ZLib zlib;
+  zlib.SetGzipHeaderMode();
+  uLongf result_size = ZLib::MinCompressbufSize(decoded.length());
+  std::unique_ptr<char[]> result(new char[result_size]);
```

Comment:
Similarly, you can use Firebase UniquePtr as well

Reply:
Done